File: ts/js/algos/searching/BinarySearch.js
"use strict";
/**
 * Binary Search
 *
 * Time Complexity: O(log(n)) - since we split search area by two for every next iteration.
 *
 * @param arr - An array of numbers/strings
 * @param target - A number/string
 * @returns Index of target
 */
Object.defineProperty(exports, "__esModule", { value: true });
function BinarySearch(arr, target) {
    let lo = 0, hi = arr.length - 1;
    while (lo <= hi) {
        let mid = lo + Math.floor((hi - lo) / 2);
        if (arr[mid] === target)
            return mid;
        if (arr[mid] > target)
            hi = mid - 1;
        if (arr[mid] < target)
            lo = mid + 1;
    }
    return -1;
}
exports.default = BinarySearch;
var a = [4, 3, 2, 3, 4, 5, 5, 5, 3, 3];
console.log(BinarySearch(a, 3), BinarySearch(a, 22));

File: algos/searching/BinarySearch.py
def binary_search(arr, l, r, key):
    if (l > r):
        return -1
    mid = l + (r - l) // 2
    if(arr[mid] == key):
        return mid
    if(arr[mid] > key):
        return binary_search(arr, l, mid - 1, key)
    else:
        return binary_search(arr, mid + 1, r, key)


def binary_search_iterative(arr, l, r, key):
    while(l <= r):
        mid = (l + r) // 2
        if(arr[mid] == key):
            return mid
        if(arr[mid] < key):
            l = mid + 1
        else:
            r = mid -1

    return -1
        

def search(arr, key):
    return binary_search(arr, 0, len(arr) - 1, key)


if __name__ == "__main__":
    arr = [2, 3, 4, 1, 6, 7, 3]
    key = 72
    print(f"{arr} {key} ==> ", search(arr, key) + 1)